File: poelevelinghelper/src/main/java/com/wittycode/poelevelinghelper/buildimport/services/domain/PobToXmlConverter.java
package com.wittycode.poelevelinghelper.buildimport.services.domain;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Base64;
import java.util.zip.DataFormatException;
import java.util.zip.Inflater;

import org.springframework.stereotype.Service;

/**
 * converts given content from PoB "raw" pastebin to XmlTree represented as String
 */
@Service
public class PobToXmlConverter {

    public String convertPobRawToXmlString(final String pobRaw) {
        final byte[] decodedPastbin = Base64.getUrlDecoder().decode(pobRaw);

        final Inflater inflater = new Inflater();
        inflater.setInput(decodedPastbin);

        byte[] buffer = new byte[1024];

        String decompressedXmlFromPasteBin = "";

        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream(decodedPastbin.length)) { 
            while (!inflater.finished()) {  
                int count = inflater.inflate(buffer);
                outputStream.write(buffer, 0, count);
            }
            decompressedXmlFromPasteBin = new String(outputStream.toByteArray());
        } catch (final IOException | DataFormatException exception) {
            System.err.println(exception.getMessage());
        }

        return decompressedXmlFromPasteBin;
    }
}

File: src/main/kotlin/com/wittycode/poelevelinghelper/buildimport/services/domain/PobToXmlConverter.kt
package com.wittycode.poelevelinghelper.buildimport.services.domain

import org.springframework.stereotype.Service
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.*
import java.util.zip.DataFormatException
import java.util.zip.Inflater

@Service
class PobToXmlConverter {

    fun convertPobRawToXmlString(pobRaw: String?): String? {
        val decodedPastbin = Base64.getUrlDecoder().decode(pobRaw)
        val inflater = Inflater()
        inflater.setInput(decodedPastbin)
        val buffer = ByteArray(1024)
        var decompressedXmlFromPasteBin: String? = ""
        try {
            ByteArrayOutputStream(decodedPastbin.size).use { outputStream ->
                while (!inflater.finished()) {
                    val count = inflater.inflate(buffer)
                    outputStream.write(buffer, 0, count)
                }
                decompressedXmlFromPasteBin = String(outputStream.toByteArray())
            }
        } catch (exception: IOException) {
            System.err.println(exception.message)
        } catch (exception: DataFormatException) {
            System.err.println(exception.message)
        }
        return decompressedXmlFromPasteBin
    }
}